File: Java/Fundamentals/Language/OOP/InterfaceTest2.java
interface TaxPayer{
	
	double getAnnualIncome();

	default double getIncomeTax(int age){
		float rate = age < 60 ? 0.15f : 0.12f;
		double income = getAnnualIncome();
		return income > 120000 ? rate * (income - 120000) : 0;
	}
}

class Consultant extends payroll.Employee implements TaxPayer{
	
	public Consultant(int h, float r){
		super(h, r);
	}

	public double getAnnualIncome(){
		return 12 * super.getIncome() + 25000;
	}
}

class Dealer implements TaxPayer{
	
	private double sales;

	public Dealer(double s){
		sales = s;
	}

	public double getAnnualIncome(){
		return 0.15 * sales;
	}
}

class InterfaceTest2{

	public static void main(String[] args){
		TaxPayer jill = new Consultant(200, 105);
		TaxPayer jack = new Dealer(5800000);
		System.out.printf("Income tax of Jill the consultant: %.2f%n", jill.getIncomeTax(36));
		System.out.printf("Income tax of Jack the dealer    : %.2f%n", jack.getIncomeTax(63));
	}
}

File: MS.NET/Fundamentals/Language/OOP/interfacetest2.cs
using System;

interface ITaxPayer
{
	double GetAnnualIncome();
}

static class TaxPayer
{
	//extension method for ITaxPayer (this qualified first parameter type)
	public static double GetIncomeTax(this ITaxPayer that, int age)
	{
		float rate = age < 60 ? 0.15f : 0.12f;
		double extra = that.GetAnnualIncome() - 120000;

		return extra > 0 ? rate * extra : 0;
	}
}

class Consultant : Payroll.Employee, ITaxPayer
{
	public Consultant(int h, float r) : base(h, r){}

	public double GetAnnualIncome()
	{
		return 12 * base.GetIncome();
	}
}

class Dealer : ITaxPayer
{
	private double sales;

	public Dealer(double s)
	{
		sales = s;
	}

	public double GetAnnualIncome()
	{
		return 0.15 * sales;
	}
}

static class Program
{
	public static void Main()
	{
		ITaxPayer jill = new Consultant(190, 210);
		ITaxPayer jack = new Dealer(5800000);

		Console.WriteLine("Jill's annual income is {0:0.00} and income tax is {1:0.00}", jill.GetAnnualIncome(), jill.GetIncomeTax(36)); //TaxPayer.GetIncomeTax(jill, 36);			
		Console.WriteLine("Jack's annual income is {0:0.00} and income tax is {1:0.00}", jack.GetAnnualIncome(), jack.GetIncomeTax(63)); 
	}
}
